nomad/deploymentwatcher/deployment_watcher.go

```diff
@@ -312,100 +342,236 @@ func (w *deploymentWatcher) StopWatch() {
 	w.exitFn()
 }
 
-// watch is the long running watcher that takes actions upon allocation changes
+// watch is the long running watcher that watches for both allocation and
+// deployment changes. Its function is to create evaluations to trigger the
+// scheduler when more progress can be made, to fail the deployment if it has
+// failed and potentially rolling back the job.
 func (w *deploymentWatcher) watch() {
+	// Get the deadline. This is likely a zero time to begin with but we need to
+	// handle the case that the deployment has already progressed and we are now
+	// just starting to watch it.
+	currentDeadline := getDeploymentProgressCutoff(w.getDeployment())
+	var deadlineTimer *time.Timer
+	if currentDeadline.IsZero() {
+		deadlineTimer = time.NewTimer(0)
+		if !deadlineTimer.Stop() {
+			<-deadlineTimer.C
+		}
+	} else {
+		deadlineTimer = time.NewTimer(currentDeadline.Sub(time.Now()))
+	}
+
 	allocIndex := uint64(1)
+	var updates *allocUpdates
+
+	rollback, deadlineHit := false, false
+
+FAIL:
 	for {
-		// Block getting all allocations that are part of the deployment using
-		// the last evaluation index. This will have us block waiting for
-		// something to change past what the scheduler has evaluated.
-		allocs, index, err := w.getAllocs(allocIndex)
-		if err != nil {
-			if err == context.Canceled || w.ctx.Err() == context.Canceled {
+		select {
+		case <-w.ctx.Done():
+			return
+		case <-deadlineTimer.C:
+			// We have hit the progress deadline so fail the deployment. We need
+			// to determine whether we should rollback the job by inspecting
```

Comment:
rollback -> roll back

Reply:
I believe rollback one word is correct: https://en.oxforddictionaries.com/definition/us/rollback